### ISSUE TITLE
Model output is not the same after speedup.

### ISSUE THREAD

**Issue by nikitakaraevv** [microsoft/nni#3010]
**Environment**:
- NNI version: commit f1b8cd2cf627edefeeb0f316a6b9e47236557530
- NNI mode (local|remote|pai): remote
- Client OS: Ubuntu 16.04.5 LTS (Xenial Xerus)
- Python version: Python 3.6.4
- PyTorch/TensorFlow version: PyTorch 1.5.0
- Is conda/virtualenv/venv used?: pyenv
- Is running in Docker?: No

**Log message**:
 - nnimanager.log: 
```
INFO:nni.compression.torch.pruning.one_shot:Pruning the dependent layers: layer3.0.conv1
INFO:nni.compression.torch.pruning.one_shot:Pruning the dependent layers: layer3.0.conv2
INFO:nni.compression.torch.pruning.one_shot:Pruning the dependent layers: layer3.2.conv3,layer3.1.conv3,layer3.0.downsample.0,layer3.5.conv3,layer3.0.conv3,layer3.3.conv3,layer3.4.conv3
INFO:nni.compression.torch.pruning.one_shot:Pruning the dependent layers: layer3.1.conv1
INFO:nni.compression.torch.pruning.one_shot:Pruning the dependent layers: layer3.1.conv2
INFO:nni.compression.torch.pruning.one_shot:Pruning the dependent layers: layer3.2.conv1
INFO:nni.compression.torch.pruning.one_shot:Pruning the dependent layers: layer3.2.conv2
INFO:nni.compression.torch.pruning.one_shot:Pruning the dependent layers: layer3.3.conv1
INFO:nni.compression.torch.pruning.one_shot:Pruning the dependent layers: layer3.3.conv2
INFO:nni.compression.torch.pruning.one_shot:Pruning the dependent layers: layer3.4.conv1
INFO:nni.compression.torch.pruning.one_shot:Pruning the dependent layers: layer3.4.conv2
```
 - dispatcher.log: N/A
 
**What issue meet, what's expected?**:
I have different outputs before and after model speedup:
<img width="657" alt="Screenshot 2020-10-21 at 11 20 25" src="https://user-images.githubusercontent.com/37815420/96693021-6cb53c80-138f-11eb-8cf9-f52a00982e00.png">

It's expected that outputs would be roughly the same.

**How to reproduce it?**: 
Please, run the code from the following gist:
https://gist.github.com/nikitakaraevv/1eaa735d5323f833356152f6d41ef8e4

**Additional information**:
I'd like to understand why it happens first of all and how it can be avoided.

**Comment by zheng-ningxin** [comment]
HI~ thanks for the feedback! It's normal that the output value changes after the speedup, basically, it is caused by the `bias` in the BN layers.

 For example, we have a conv-bn block in the model, and the conv layer and bn layer both have three output channels. Before the speedup, if you prune the first output channel of the conv layer, the first output channel of the conv layer should be all zeros. However, the following BN layer will make the first output channel of conv change back to some non-zero values(`bias` in the BN layer), and these values will be reflected in your final output. After the speedup, the first channel of the conv layer and BN layer will be actually removed, so those `non-zero values` caused by the BN will also be removed. This is the source of the differences.


To check if the speedup works fine, you can set the bias of the BN layers to zero (for example, `zero_bn_bias` in [test_model_speedup.py](https://github.com/microsoft/nni/blob/master/src/sdk/pynni/tests/test_model_speedup.py) ) and then compare whether the outputs are the same.

**Comment by nikitakaraevv** [comment]
Hi Ningxin,

Thank you for your answer!
Indeed, after setting all the BN biases to zero it works fine.

I've retrained a network after pruning and after pruning+speedup. It wasn't able to achieve the same level of quality after speedup which made me think that it is not only due to BN. 

You can see the difference below (0 corresponds to restoring from a trained checkpoint, then pruning 50% of weights, then speedup, 1 corresponds to restoring and pruning of 50% of weights)
 <img width="540" alt="Screenshot 2020-10-22 at 10 55 05" src="https://user-images.githubusercontent.com/37815420/96841920-10b4eb80-1455-11eb-9473-bb8304e1fa2c.png">

Moreover, if I do speedup after pruning+retraining, the model is no longer able to give reasonable predictions and I have to retrain it anyway.

Maybe you know if there is a way to keep BN biases or some other way to avoid this kind of behaviour after speedup? Perhaps, there is a different way of doing pruning+speedup+retraining?

**Comment by zheng-ningxin** [comment]
Hi~ Sorry for the late reply and thanks for the new feedback! 
How many filters are there in the model, and how many of them have you pruned ? 
I am not sure what caused it, but my guess is:
(1) The trainable parameters of the model before speedup are more than those after speedup (BN bias, running_std, etc)
(2) The hyperparameters you use for retraining the speeduped model may not be appropriate.
If it is convenient, could you please provide your model and training hyperparameters, so that we can dig into this problem?

**Comment by nikitakaraevv** [comment]
Hi Ningxin, 
Thank you for your response. 

I pruned 50% of weights in selected Conv filters of my backbone which makes almost 40% of all the network weights. I tried to prune different filters with different parameters. These ⬆️ are the best metrics I've got. 

Unfortunately, I can't provide the model I've trained but I will try to reproduce this experiment on another model and provide you hyperparameters I used as well as the model itself.

By the way, I've noticed that for such backbones as mobilenet v2 speed-up after both usual and dependency-aware FPGM pruning doesn't really reduce the model size since this architecture has group convolutions with `number of input channels == number of output channels == number of groups`. So, filters can't be removed due to the fact that the number of input and output channels must be divisible by the number of groups.

I've checked all of your tutorials but haven't found and example of Mobilenet speedup using one of l1filter / l2filter / fpgm pruning. Is there a way to reduce the MobileNet v2 size after pruning with one of this algorithms?

**Comment by zheng-ningxin** [comment]
@nikitakaraevv We have already supported the pruning for Mobilenet/MobilenetV2 in version 1.9. Please have a try on v1.9~ Thanks~

**Comment by twmht** [comment]
@zheng-ningxin 

do we need to disable bn bias when training with like slim pruner?

**Comment by zheng-ningxin** [comment]
> 
> 
> @zheng-ningxin
> 
> do we need to disable bn bias when training with like slim pruner?
I suggest speedup the model before finetuning so that we can avoid interference from the bias. Disable the corresponding bn bias of the pruned channel during the finetuning(with the mask) can also work, but it is difficult.

**Comment by twmht** [comment]
@zheng-ningxin 

what does "finetuning" mean here? taking slimPruner as an example (https://github.com/microsoft/nni/blob/master/examples/model_compress/slim_torch_cifar10.py#L111),  first step is to train from the scratch,  second step is to prune the model and generate the mask. and final step is to run model speedup, I don't see any fine-tuning here. is this a complete example?

**Comment by scarlett2018** [comment]
> @nikitakaraevv We have already supported the pruning for Mobilenet/MobilenetV2 in version 1.9. Please have a try on v1.9~ Thanks~

As the major issue has been addressed, closing the issue.
